### ISSUE TITLE
Live Updates and Push

### ISSUE THREAD

**Issue by bblfish** [solid/notifications#35]
The WebApp WG is working on a [Push API](http://www.w3.org/TR/push-api/). Since SoLiD has a section on [Live Updates](https://github.com/solid/solid-spec#live-updates) it would be useful to understand how these two complement each other, work together, what ideas can be taken from it, etc...

Some resources:
- the blog post [Push Notifications on the Open Web](http://updates.html5rocks.com/2015/03/push-notificatons-on-the-open-web)
- the IETF draft [Generic Event Delivery Using HTTP Push](https://datatracker.ietf.org/doc/draft-thomson-webpush-protocol/?include_text=1)

Any thoughts?

**Comment by elf-pavlik** [comment]
I was thinking about it a little bit a few years ago: https://forum.solidproject.org/t/push-notifications-push-api/1079

I still think that this would require the application (OAuth client) to have a server-side component. This server-side component could use one of the *target* based #33 subscription types to subscribe to storage.
From there it would be pretty much that application implementation detail how Push API is being used.

Especially given Push API being a big part of what drives UX of the app, I think how it's being used should be fully up to the application developer.

In other words, Push API would be an implementation detail of each application using it. I don't think we need to specify anything directly related to it. What we need:

* *target* based subscription types which server-side component of the application would use.
* run use case of application (OAuth clients) which have both server-side and on-device components through AuthN and AuthZ panels.  I think the End-user shouldn't be burdened with the detail of certain clients having two components so bother components should use the same Client ID and be considered the same application.

**Comment by uvdsl** [comment]
Hi all,

I built a demo for Web Push Notifications from Solid Pods.

The demo consists of three components:
- a Solid Pod (Server)
- which is extended with a [Web Push module](https://github.com/uvdsl/solid-web-push/)
- such that my [Progressive Web App](https://github.com/uvdsl/solid-web-pwa/) can have Web Push Notifications

In addition to the readme at the [Web Push module](https://github.com/uvdsl/solid-web-push/),
we have a supplementary [demo website](https://uvdsl.solid.aifb.kit.edu/conf/2022/icwe/demo) with a demo walkthrough, a sequence diagram, the data model and a [demo video](https://uvdsl.solid.aifb.kit.edu/conf/2022/icwe/demo#v).

How it works in a nutshell:
- The Web Push module has an inbox (at the Pod) where subscriptions can be posted to by a PWA.
- These subscriptions contain the target resource, the PWA would like receive Push Notifications on.
- Upon a new subscription arriving in its inbox, the Web Push module processes the subscription and watches the target resource for updates.
- When a target resource is updated, a Web Push notification is send to the PWAs that are currently subscribed to the target resource.

I would like to point out:
- Providing Web Push out-of-the-box with Solid Pods 
	- may facilitate development of client-side applications that do not rely on server-side components.
	- may simplify authentication and authorization as we do not need some special OAuth Token (e.g., scope "notification") from some third party backend / server as the Pod could execute this check upon submission of the subscription to the service inbox.
	- does not forbid to have server-side components. Instead, Web Push is already there to be leveraged via server-to-server (or Pod-to-Pod) notification schemes.

- An application developer 
	- _does not need_ to use Solid Web Push. He can implemnt his/her own service.
	- can rely on Solid Web Push to forward notifications that are sent via other Notification Schemes to the user's Pod.
	- can set up an application-specific inbox that he can post notifications (sales,re-engagement,...) to, which will be forwarded to the client by Solid Web Push. 
    		- So the UX is still up to the app developer.
    		- It's easy: Set up app-inbox, subscribe to the app inbox, have Web Push Notifications, profit!

- Caveat:
	- As far as I know, a PWA (or more specific: a service worker) can only have one subscription to a push service. To receive Push Notifications from resources distributed across multiple Pods, the application developer would need to combine Pod-to-Pod notifications which are then forwarded by the Web Push service living on the user's Pod (server). This may seem complicated, but I deem it less complicated than to create a whole server-side component for one's app for that purpose.

I would like to emphasise that I do not want to take anything away from the currently proposed Solid Notification Spec.

**Comment by elf-pavlik** [comment]
I wonder if `userVisibleOnly: true`, which is required for example in Chrome, also comes as a certain Caveat? In many cases, applications may want to receive updates but don't show a notification to the user every time some resource(s) change.

**Comment by uvdsl** [comment]
The flag `userVisibleOnly` mandates that the user has to be able to recognise when the app receives a push message from the server.
Practically, this means that if the app is closed or in background, a notification must be displayed to the user when a push message is received. (The design of that notification is customisable by the application developer.)
When the app is in foreground / active, this does not apply. In this case, the app can "silently" process the push message from the server, e.g. to update the UI. 

In my demo PWA, this is the case when you have enabled Web Push Notifications and are viewing the inbox.
When the inbox is updated on the Pod, no notification is displayed, even though the PWA received a push message about the updated inbox. The PWA silently updates the UI.

**Comment by elf-pavlik** [comment]
@uvdsl For the PR you are working on. Will you align it with the common subscription mechanism described in the Notifications Protocol draft or you will keep the inbox approach used in your demo?

**Comment by uvdsl** [comment]
Are you referring to the notion of a [subscription resource](https://solid.github.io/notifications/protocol#notification-subscription) that accepts a `POST` for subscription on a topic, i.e. some other resource? If so, I would like to point out that my "inbox approach" is basically the same, except it assumes that the `subscription resource` is an [LDP Container](https://www.w3.org/TR/ldp/#ldpc-container) instead of some arbitrary REST endpoint (which I assume the resource to be, albeit defined as a RDF document).

Personally, I like well-defined endpoints. And I can imagine a service offering an LDP-based API where the resource gets processed immediately. One may get a `201 Created` response with a `Location` header but that does not mean that one has access to that resource at all or if that resource has not already been processed and been deleted. So, in my simple mind, there is virtually no difference here, except that the API is a little better defined.

If there has not been a hard-set decision yet on the design of the subscription resource, then I would like to offer the LDP take on it for discussion and thus keep it in my draft for now.

Regarding the data model or the body of the `POST` request, I would be happy to adjust my modelling as suggested by more experienced community members if that helps smoothing out differences.

If I am not mistaken, @csarven had an opinion on the inbox/LDP approach, though I do not recall if he liked it or not :)

**Comment by uvdsl** [comment]
Looking twice, I see the difference now, between the RDF document describing the subscription resource. So one can get a representation of that resource with `GET`. Okay, that seems fair. 
But still, the definition of `POST` on that resource... Maybe it is worth discussing this in a panel meeting? What do you think, @elf-pavlik ?

**Comment by uvdsl** [comment]
Oh, I just noticed there exists already an issue on that (https://github.com/solid/notifications/issues/36)!
Moving discussion on that specific topic there.
